master/pkg/ws/ws.go

```diff
@@ -0,0 +1,244 @@
+				return fmt.Errorf("message size %d exceeds maximum size %d", len(bs), maxMessageSize)
+			}
+
+			switch err := s.conn.WriteMessage(websocket.TextMessage, bs); {
+			case err == websocket.ErrCloseSent:
+				return nil
+			case err != nil:
+				return fmt.Errorf("writing message: %w", err)
+			}
+		case <-ping.C:
+			err := s.conn.WriteControl(websocket.PingMessage, nil, time.Now().Add(pongWait))
+			netErr, ok := err.(net.Error)
+			switch {
+			case ok && netErr.Timeout():
+				continue
+			case err == websocket.ErrCloseSent:
+				return nil
+			case err != nil:
+				return fmt.Errorf("sending ping: %w", err)
+			}
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+// Close closes the websocket by performing the close handshake and closing the underlying
+// connection, rendering it unusable.
+func (s *Websocket[TIn, TOut]) Close() error {
+	s.closeOnce.Do(func() {
+		s.log.Trace("attempting graceful close")
+		if hErr := s.closeGraceful(); hErr != nil {
+			s.setError(fmt.Errorf("gracefully closing: %w", hErr))
+			s.log.Trace("attempting forceful close")
+			if fErr := s.closeForced(); fErr != nil {
+				s.log.WithError(fErr).Error("failed to forcibly close socket")
+			}
+		}
+		s.log.Trace("socket closed")
+	})
+	return s.Error()
+}
+
+func (s *Websocket[TIn, TOut]) closeGraceful() error {
+	s.cancel()
+
+	closeDeadline := time.Now().Add(closeWait)
+	s.conn.SetPongHandler(nil) // So the pong handler doesn't extend the deadline.
+	if err := s.conn.SetReadDeadline(closeDeadline); err != nil {
+		return fmt.Errorf("setting read deadline: %w", err)
+	}
+
+	// If this close message begins the handshake, the read loop will exhaust messages until our
+	// peer responds with their close, or it exceeds the read deadline (or, you know, the underlying
+	// connection is ripped from its hands), then exit. If we are already closed (have received and
+	// responded with, by the default close handler, a close), we will receive ErrCloseSent from the
+	// write and the read loop should have already exited.
+	if err := s.conn.WriteControl(
+		websocket.CloseMessage,
+		websocket.FormatCloseMessage(websocket.CloseNormalClosure, "close called"),
+		closeDeadline,
+	); err != websocket.ErrCloseSent && err != nil {
+		return fmt.Errorf("sending close: %w", err)
+	}
+
+	<-s.Done
+	if clErr := s.conn.Close(); clErr != nil {
+		return fmt.Errorf("closing underlying conn: %w", clErr)
+	}
+	return nil
+}
+
+func (s *Websocket[TIn, TOut]) closeForced() error {
+	s.cancel()
+	if err := s.conn.Close(); err != nil {
+		<-s.Done
+		return fmt.Errorf("closing underlying conn: %w", err)
+	}
+	<-s.Done
+	return nil
+}
+
+// Error returns an error if the Websocket has encountered one. Errors from closing are excluded.
+func (s *Websocket[TIn, TOut]) Error() error {
+	s.errLock.Lock()
+	defer s.errLock.Unlock()
+	return s.err
+}
+
+func (s *Websocket[TIn, TOut]) setError(err error) {
+	s.errLock.Lock()
+	defer s.errLock.Unlock()
+	if err == nil {
```

Comment:
```suggestion
	if err != nil {
```
Or just remove the check?

And I guess consider figuring out if you can make a test that causes and checks for an error.

Reply:
The point is just that the first error wins. I can invert the flow, maybe clearer.

Reply:
Sounds like you mean for this to be `s.err == nil`, which would also be fine. But right now it has `err` rather than `s.err`.

Reply:
fixed. will add failure case tests soon-ish.